src/clj/puppetlabs/services/jruby_pool_manager/impl/instance_pool.clj

```diff
@@ -0,0 +1,62 @@
+(ns puppetlabs.services.jruby-pool-manager.impl.instance-pool
+  (:require [puppetlabs.services.jruby-pool-manager.impl.jruby-agents :as jruby-agents]
+            [puppetlabs.services.protocols.jruby-pool :as pool-protocol]
+            [puppetlabs.services.jruby-pool-manager.impl.jruby-internal :as jruby-internal]
+            [puppetlabs.services.jruby-pool-manager.jruby-schemas :as jruby-schemas]
+            [clojure.tools.logging :as log]
+            [puppetlabs.i18n.core :as i18n])
+  (:import (puppetlabs.services.jruby_pool_manager.jruby_schemas InstancePool)))
+
+(extend-type InstancePool
+  pool-protocol/JRubyPool
+  (fill
+    [this]
+    (let [modify-instance-agent (jruby-agents/get-modify-instance-agent this)]
+      (jruby-agents/send-agent modify-instance-agent
+                               #(jruby-agents/prime-pool! this))))
+
+  (shutdown
+    [this]
+    (jruby-agents/flush-pool-for-shutdown! this))
+
+  (lock
+    [this]
+    (let [pool (jruby-internal/get-pool this)]
+      (.lock pool)))
+
+  (lock-with-timeout
+    [this timeout time-unit]
+    (let [pool (jruby-internal/get-pool this)]
+      (.lockWithTimeout pool timeout time-unit)))
+
+  (unlock
+    [this]
+    (let [pool (jruby-internal/get-pool this)]
+      (.unlock pool)))
+
+  (borrow
```

Comment:
This currently isn't being called anywhere. There is (currently) no difference in borrow behavior between the two implementations. Also a lot of tests (in both this lib and in puppetserver) call `borrowItem` directly on the pool. But for consistency, this and its sister function `borrow-with-timeout` should probably be properly implemented and used.